File: BOJ/java/1759.java
import java.io.IOException;
import java.util.Arrays;
import java.util.Scanner;

class Main {
    static int l, c;
    static String[] a, t;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        l = sc.nextInt();
        c = sc.nextInt();
        sc.nextLine();
        a = sc.nextLine().split(" ");
        t = new String[c];
        Arrays.sort(a);

        dfs(0, 0);
        System.out.println(sb);
    }

    static void dfs(int at, int depth) {
        if (depth == l && isPossible(t)) {
            for (int i = 0; i < l; i++) {
                sb.append(t[i]);
            }
            sb.append("\n");
            return;
        }

        for (int i = at; i < c; i++) {
            t[depth] = a[i];
            dfs(i + 1, depth + 1);
        }
    }

    static boolean isPossible(String[] s) {
        int vowel = 0, consonant = 0;
        for (int i = 0; i < l; i++) {
            if (isVowel(s[i])) vowel++;
            else consonant++;
        }
        return vowel >= 1 && consonant >= 2;
    }

    static boolean isVowel(String s) {
        char a = s.charAt(0);
        return a == 'a' || a == 'e' || a == 'i' || a == 'o' || a == 'u';
    }
}

File: BOJ/kotlin/1759.kt
import java.util.*

fun main() = with(Scanner(System.`in`)) {
    val l = nextInt()
    val c = nextInt()
    val str = Array(c) { next()[0] }.sortedArray()

    val used = Array(c) { false }

    fun find(i: Int, n: Int) {
        used[i] = true

        if (n < l) {
            for (x in i + 1 until c) find(x, n + 1)
        } else {
            val pass = (0 until c)
                .filter { used[it] }
                .map { str[it] }
                .joinToString("")

            if (pass.count { it in "aeiou" } >= 1 &&
                pass.count { it !in "aeiou" } >= 2) {
                println(pass)
            }
        }

        used[i] = false
    }

    for (i in 0 until c) find(i, 1)
}